src/traces/main.cpp

```diff
@@ -0,0 +1,164 @@
+#include <iostream>
+#include <random>
+#include <vector>
+
+#include <time.h>
+
+#include "../common/logger.hpp"
+#include "../common/status.hpp"
+#include "../common/types.hpp"
+
+#include "../sequential/sequential.hpp"
+
+#include "../parallel/parallel.hpp"
+
+#include <iostream>
+#include <optional> // std::optional i think is c++17
+#include <pthread.h>
+
+// structure to pass data to threads
+struct OperationData {
+    enum OperationType { INSERT, SEARCH, DELETE };
+
+    OperationType operation;
+    KeyType key;
+    ValueType value;
+};
+
+// Define the structure to pass data to threads
+struct ThreadData {
+    ParallelRobinHoodHashTable *hashTable;
+    OperationData *operations;
+    size_t numOperations;
+
+    int threadID;
+};
```

Comment:
This could all be passed into the C++ thread initializer. And you can use a `std::vector` too!

```c++
auto run_trace = [] (ParallelRobinHoodHashTable &hash_table, std::vector<OperationData> &trace, int t_id) {
    ...
};
std::thread worker(run_trace, hash_table, trace, t_id);
```

Disclaimer: idk if this is entirely correct, but hopefully you get the gist

Reply:
You don't even need to use the lambda function `auto run_trace = ...`. But it would be much cleaner if you did it the C++ way.

For multiple threads, try:

```c++
std::vector<std::thread> workers;

for (int worker_id = 0; worker_id < NUM_WORKERS; ++worker_id) {
    workers.emplace_back(run_trace, hash_table, trace, worker_id);
}

for (int worker_id = 0; worker_id < NUM_WORKERS; ++worker_id) {
    workers[worker_id].join();
}
```

Reply:
so run trace is a function?